File: projects/grind-75/python/graph/word_ladder.py
from collections import defaultdict, deque
from typing import List


# TODO: Bi-directional BFS?
class Solution:
    @classmethod
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0

        graph = defaultdict(list)
        for word in wordList + [beginWord]:
            for i in range(len(word)):
                graph[word[0:i] + "*" + word[i + 1 :]].append(word)

        q = deque([beginWord])
        visited = set()
        res = 0
        while q:
            res += 1
            for _ in range(len(q)):
                word = q.popleft()
                if word == endWord:
                    return res
                for i in range(len(word)):
                    key = word[0:i] + "*" + word[i + 1 :]
                    if key not in visited:
                        q += graph[key]
                        visited.add(key)
        return 0

File: projects/grind-75/typescript/graph/word-ladder.ts
export function ladderLength(beginWord: string, endWord: string, wordList: string[]): number {
    if (!wordList.includes(endWord)) return 0;
    const graph: Record<string, string[]> = {};
    const addToGraph = (key: string, word: string) => {
        if (!(key in graph)) {
            graph[key] = [];
        }
        graph[key].push(word);
    };

    for (const word of wordList.concat(beginWord)) {
        for (let i = 0; i < word.length; i++) {
            const key = word.slice(0, i) + "*" + word.slice(i + 1);
            addToGraph(key, word);
        }
    }

    const q: string[] = [beginWord];
    const visited: Set<string> = new Set();
    let res = 0;
    while (q.length) {
        res += 1;
        const levelSize = q.length;
        for (let i = 0; i < levelSize; i++) {
            const word = q.shift()!;
            if (word === endWord) {
                return res;
            }

            for (let j = 0; j < word.length; j++) {
                const key = word.slice(0, j) + "*" + word.slice(j + 1);
                if (!visited.has(key)) {
                    if (key in graph) {
                        q.push(...graph[key]);
                    }
                    visited.add(key);
                }
            }
        }
    }

    return 0;
}